Attempt to make compatible with new Parsetree interface. Tests pass

rdavison opened the pull request:
Forgive me if I'm doing something wrong here, this is my first PR to an ocaml library.  I was trying to experiment with this library when I discovered that it is incompatible with 4.03.0 due to some changes in Parsetree.  I made the changes locally, reran the tests, and everything seems to pass.  I was able to build the examples as well.  

It appears that a new feature was added to parse integers with suffixes, as you can see here: https://github.com/ocaml/ocaml/blob/trunk/parsing/parsetree.mli#L22

My changes don't attempt to handle the suffixes (maybe they should?), but I believe that it is able to handle the original untyped case.  If there is anything I can do to help resolve this issue, I would be glad to help out.

nojb commented:
Hi!  Thanks very much for taking the time to contribute.  I was actually meaning to bring this up to scratch and make a release.  The initial porting process (from the old `camlp4` bitstring syntax extension) is mostly done.  I just haven't had the time to think through the packaging issues.  I remember I was a little bit blocked trying to make it compatible with the old `bitstring` ocamlfind package.  In any case, I do not think this is necessary anymore as `camlp4` is mostly deprecated.

I will look at your patch (I think it is fine) and merge it tonight.  If you would like to work / help release / help mantain this project I would gladly accept any help and could provide guidance, etc.

rdavison commented:
Oh, yes, I noticed that the `bitstring` code in the repo is different from the original `bitstring` package. However, it seems that the original package has not changed in many years.

Regarding, the packaging, it did seem a bit strange, but I'm mostly new to ocaml so I haven't yet learned much about packaging.  Also, maybe you can help me with this.  I was blocked on getting the ppx extension to actually run through the regular channels.  Eventually I found that if I invoke ocamlfind with an absolute path to `ppx_bitstring.byte` then it would work, like this:

```ocamlfind ocamlc -o elf.out -ppx /home/rdavison/.opam/4.03.0/lib/ppx_bitstring/ppx_bitstring.byte -linkpkg -package ppx_bitstring elf.ml```

But it seems strange to me that I have to include the entire path instead of some shortcut.  Is there something I'm missing out on here that you might be aware of?

nojb commented on src/bitstring_persistent.ml:
```diff
@@ -180,7 +180,7 @@ let named_from_string = Marshal.from_string
   {
     field = Pat.any ();
-    flen = Exp.constant (Const_int 32);
+    flen = Exp.constant (Pconst_integer ("32", None));
```
Here, and below, you can use the functions from `Ast_helper.Const` (see https://github.com/ocaml/ocaml/blob/trunk/parsing/ast_helper.mli#L38-L47).

nojb commented:
By the way, it would be great if you could check if any changes are needed for 4.04.0, which was recently released.

nojb commented:
I renamed the repository/project `bitstring-ppx` since there already exist opam packages `bitstring` and `ppx_bitstring`.  If you pull the latest changes, you can invoke the `ppx` extension by doing

    ocamlfind ocamlc -o elf.out -package bitstring-ppx.ppx -linkpkg elf.ml

(assuming you have installed with `opam`)

rdavison commented:
Awesome, thanks for the updates!  So, I updated my repo with the upstream changes and switched to OCaml 4.04.0.  The functions in the Const module worked great.  The only question I think I have at the moment is in regards to this commit I made at the last moment: https://github.com/rdavison/bitstring-ppx/commit/874d65eaac5176c8b072e3428c4f373976676083

I decided to get rid of the wildcard match because it was actually a shortcut that I took because I'm not certain what the behavior should be whenever it encounters annotated integers.  It seems alright to me that they can be ignored, at least the non-standard ones.  Do you have any other insight?

rdavison commented:
Ok, here is the updated version: https://github.com/rdavison/bitstring-ppx/commit/f72cfc1fc007618cf1ab0053ffb8c4abbef55cb2

nojb commented:
Looks good, will try it out tonight or tomorrow morning and merge it if there are no other issues.

BTW, it seems that the email you are committing with (`git config user.email`) is not registered with Github so your commits are not being linked with your account.  

Thanks!

rdavison commented:
Oh, good catch, I didn't notice! I've updated my email address in github so it reflects the address in the commit history. Thanks

nojb commented:
Merged! Thanks a lot for your contribution!

I will submit an opam release soon.  However, I am still not convinced that the ppx syntax used is the best.  Maybe will do a "beta" release and ask for suggestions before doing a full release.

nojb commented:
Also, there are a couple of issues (https://github.com/nojb/bitstring-ppx/issues) left open from some time back - but I am not sure I have time right now to look at them ... If you want to try your hand at them, it would be a really good way to learn how things work and a great contribution !
